internal/jimm/controller.go

```diff
@@ -393,133 +393,231 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
 	// credential is not strictly necessary, but will help prevent the user thinking they can create new
 	// models on the incoming cloud.
-	allCredentials, err := j.Database.GetIdentityCloudCredentials(ctx, &ownerUser, cloudTag.Id())
+	allCredentials, err := m.jimm.Database.GetIdentityCloudCredentials(ctx, &m.model.Owner, cloudTag.Id())
 	if err != nil {
-		return errors.E(op, err)
+		return err
 	}
 	if len(allCredentials) == 0 {
-		return errors.E(op, errors.CodeNotFound, fmt.Sprintf("Failed to find cloud credential for user %s on cloud %s", ownerUser.Name, cloudTag.Id()))
+		return errors.E(errors.CodeNotFound, fmt.Sprintf("Failed to find cloud credential for user %s on cloud %s", m.model.Owner.Name, cloudTag.Id()))
 	}
 	cloudCredential := allCredentials[0]
 
-	model.CloudCredentialID = cloudCredential.ID
-	model.CloudCredential = cloudCredential
+	m.model.CloudCredentialID = cloudCredential.ID
+	m.model.CloudCredential = cloudCredential
+
+	return nil
+}
 
+func (m *modelImporter) setModelCloud(ctx context.Context) error {
 	// fetch the cloud used by the model
-	cloud := dbmodel.Cloud{
-		Name: cloudCredential.CloudName,
+	cloudTag, err := names.ParseCloudTag(m.modelInfo.CloudTag)
+	if err != nil {
+		return err
 	}
-	err = j.Database.GetCloud(ctx, &cloud)
+	cloud := dbmodel.Cloud{Name: cloudTag.Id()}
+	err = m.jimm.Database.GetCloud(ctx, &cloud)
 	if err != nil {
 		zapctx.Error(ctx, "failed to get cloud", zap.String("cloud", cloud.Name))
-		return errors.E(op, err)
+		return err
 	}
 
-	cr := cloud.Region(modelInfo.CloudRegion)
-	if cr.Name != modelInfo.CloudRegion {
-		return errors.E(op, "cloud region not found")
+	cr := cloud.Region(m.modelInfo.CloudRegion)
+	if cr.Name != m.modelInfo.CloudRegion {
+		return errors.E("cloud region not found")
 	}
 
-	model.CloudRegionID = cr.ID
-	model.CloudRegion = cr
+	m.model.CloudRegionID = cr.ID
+	m.model.CloudRegion = cr
 
-	err = j.Database.AddModel(ctx, &model)
-	if err != nil {
-		if errors.ErrorCode(err) == errors.CodeAlreadyExists {
-			return errors.E(op, err, "model already exists")
+	return nil
+}
+
+func (m *modelImporter) save(ctx context.Context) error {
+	return m.jimm.Database.Transaction(func(d *db.Database) error {
+		err := m.jimm.Database.AddModel(ctx, &m.model)
+		if err != nil {
+			if errors.ErrorCode(err) == errors.CodeAlreadyExists {
+				return fmt.Errorf("model (%s) already exists", m.model.Name)
+			}
+			return err
 		}
+		for _, offer := range m.offersToAdd {
+			var dbOffer dbmodel.ApplicationOffer
+			dbOffer.FromJujuApplicationOfferAdminDetailsV5(offer)
+			dbOffer.ModelID = m.model.ID
+			if err := m.jimm.Database.AddApplicationOffer(ctx, &dbOffer); err != nil {
+				if errors.ErrorCode(err) == errors.CodeAlreadyExists {
+					return fmt.Errorf("offer with URL %s already exists", offer.OfferURL)
+				}
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// ImportModel imports model with the specified UUID from the controller.
```

Comment:
imports a model and the currently existing offers into JIMM. A new owner must be set to represent the external user who will own this model.

Something like this for the comment?